provider.go

```diff
@@ -1052,55 +1072,346 @@ func (p *provider) Delete(ctx context.Context, req *rpc.DeleteRequest) (*emptypb
 
 }
 
+// ConstructRequest captures enough data to be able to register nested components against the caller's resource
+// monitor.
 type ConstructRequest struct {
-	URN       presource.URN
-	Preview   bool
-	Construct func(context.Context, ConstructFunc) (ConstructResponse, error)
+	Urn     presource.URN // the Pulumi URN for this resource.
+	Info    ConstructInfo
+	Parent  presource.URN
+	Inputs  presource.PropertyMap
+	Options ConstructOptions
 }
 
-type ConstructFunc = func(
-	*pulumi.Context, comProvider.ConstructInputs, pulumi.ResourceOption,
-) (pulumi.ComponentResource, error)
+// ConstructInfo contains all of the information required to register resources as part of a call to Construct.
+type ConstructInfo struct {
+	Config           map[pconfig.Key]string // the configuration variables to apply before running.
+	ConfigSecretKeys []pconfig.Key          // the configuration keys that have secret values.
+	DryRun           bool                   // true if we are performing a dry-run (preview).
+	Parallel         int32                  // the degree of parallelism for resource operations (<=1 for serial).
+	MonitorEndpoint  string                 // the RPC address to the host resource monitor.
+}
 
-type ConstructResponse struct{ inner *rpc.ConstructResponse }
+// ConstructOptions captures options for a call to Construct.
+type ConstructOptions struct {
+	// Aliases is the set of aliases for the component.
+	Aliases []presource.URN
 
-func (p *provider) Construct(ctx context.Context, req *rpc.ConstructRequest) (*rpc.ConstructResponse, error) {
-	// This returns the URN of the parent, we just need the type.
-	parent := tokens.Type(req.GetParent())
-	if parent != "" {
-		parent = presource.URN(parent).Type()
+	// Dependencies is the list of resources this component depends on.
+	Dependencies []presource.URN
+
+	// Protect is true if the component is protected.
+	Protect bool
```

Comment:
While we're making break changes this and the other `bool` options should be `*bool` https://github.com/pulumi/pulumi/issues/18935